src/triples/batch_random_ot.rs

```diff
@@ -109,10 +185,115 @@ pub async fn batch_random_ot_receiver<C: CSCurve>(
+            ));
+        }
+    
+        let delta = BitVector::random(&mut OsRng);
+        big_yv.push(big_y);
+        deltav.push(delta);
+    }
+    
+    let big_y_v_arc = Arc::new(big_yv);
+    let big_y_affine_v_arc = Arc::new(big_y_affine_v);
+    
+    // inner is batch, outer is bits
+    let mut choices: Vec<Vec<_>> = Vec::new();
+    for _ in deltav[0].bits() {
+        choices.push(Vec::new());
+    }
+    for j in 0..N {
+        for (i, d_i) in deltav[j].bits().enumerate() {
+            choices[i].push(d_i);
+        }
+    }
+    // wrap in arc
+    let choices: Vec<_> = choices.into_iter().map(|c| Arc::new(c)).collect();
+    
+    let mut tasks = Vec::new();
+    for i in 0..choices.len() {
+        let mut chan = chan.child(i as u64);
+        // clone arcs
+        let d_i_v = choices[i].clone();
+        let big_y_v_arc = big_y_v_arc.clone();
+        let big_y_affine_v_arc= big_y_affine_v_arc.clone();
+        let task = ctx.spawn(async move {
+            let mut x_i_v = Vec::new();
+            let mut big_x_iv = Vec::new();
+            for j in 0..N {
+                let d_i = d_i_v[j];
+                // Step 4
+                let x_i = C::Scalar::random(&mut OsRng);
+                let mut big_x_i = C::ProjectivePoint::generator() * x_i;
+                big_x_i.conditional_assign(&(big_x_i + big_y_v_arc[j]), d_i);
+                x_i_v.push(x_i);
+                big_x_iv.push(big_x_i);
+            }
+            // Step 6
+            let mut big_x_i_affine_v = Vec::new();
+            let wait0 = chan.next_waitpoint();
+            for j in 0..N {
+                let big_x_i_affine = SerializablePoint::<C>::from_projective(&big_x_iv[j]);
+                big_x_i_affine_v.push(big_x_i_affine);
+            }
+            chan.send(wait0, &big_x_i_affine_v).await;
+            
+            // Step 5
+            let mut hashv = Vec::new();
+            for j in 0..N {
+                let big_x_i_affine = big_x_i_affine_v[j];
+                let big_y_affine = big_y_affine_v_arc[j];
+                let big_y = big_y_v_arc[j];
+                let x_i = x_i_v[j];
+                hashv.push(hash(i, &big_x_i_affine, &big_y_affine, &(big_y * x_i)));
+            }
+            hashv
+        });
+        tasks.push(task)
+    }
+    
+    let outs: Vec<Vec<_>> = stream::iter(tasks).then(|t| t).collect().await;
+    // batch dimension is on the inside but needs to be on the outside
+    let mut reshaped_outs: Vec<Vec<_>> = Vec::new();
+    for _ in 0..N {
+        reshaped_outs.push(Vec::new());
+    }
+    for i in 0..outs.len() {
+        for j in 0..N {
+            reshaped_outs[j].push(outs[i][j]);
+        }
+    }
+    let outs = reshaped_outs;
+    let mut ret = Vec::new();
+    for j in 0..N {
+        let delta = deltav[j];
+        let out = &outs[j];
+        let big_k: BitMatrix = out.into_iter().cloned().collect();
+        let h = SquareBitMatrix::try_from(big_k);
```

Comment:
Where did SquareBitMatrix come from?  In the original code it looks like big_k is just a bitmatrix.  Why do you need to add the concept of "h" and the SquareBitMatrix?

Reply:
on line 189 when we say big_k.try_into():

`Ok((delta, big_k.try_into().unwrap()))`

this is coercing to a SquareBitMatrix. The type is hidden inside the typedef: 

`type BatchRandomOTOutputReceiver = (BitVector, SquareBitMatrix);`